### ISSUE TITLE
GFLOPs

### ISSUE THREAD

**Issue by mugenggeng** [pangzhan27/CMeRT#2]
Could you please explain how you calculated the GFLOPs? Why is the GFLOPs value I obtained through the thop library 75G? Here is my code: `flops, params = profile(model, inputs=[x.to(device) for x in data[:-4]])` print("flops", flops)
print("params", params)

**Comment by pangzhan27** [comment]
Hi
Thanks for your interest to our work. 
I think you test the GFLOPs for a batch of samples. 
The result we reported is done with batch size set to 1. 

For your information, I include the code we used as below.

```
param_size = 0
for param in model.parameters():
    param_size += param.nelement() * param.element_size()
    logger.info('Number of parameters: {}. Model Size: {:.2f} MB'.format(sum(p.numel() for p in model.parameters()), param_size/ 1024**2))

def prepare_input(resolution):
        from rekognition_online_action_detection.models.feature_head import FEATURE_SIZES
        fuse_dim = cfg.MODEL.LSTR.LONG_MEMORY_NUM_SAMPLES+cfg.MODEL.LSTR.WORK_MEMORY_NUM_SAMPLES
        rgb = torch.FloatTensor(1, fuse_dim, FEATURE_SIZES[cfg.INPUT.VISUAL_FEATURE]).to(device)
        flow = torch.FloatTensor(1, fuse_dim, FEATURE_SIZES[cfg.INPUT.MOTION_FEATURE]).to(device)
        mask = torch.FloatTensor(1, cfg.MODEL.LSTR.LONG_MEMORY_NUM_SAMPLES).to(device)
        return {'visual_inputs': rgb, 'motion_inputs': flow,  'memory_key_padding_mask': mask}

from ptflops import get_model_complexity_info
macs, params = get_model_complexity_info(model, (1, 224, 224), as_strings=True, backend='pytorch',
                                             input_constructor=prepare_input,
                                             print_per_layer_stat = True, verbose = True)
print('{:<30}  {:<8}'.format('Computational complexity: ', macs))
print('{:<30}  {:<8}'.format('Number of parameters: ', params))
```

**Comment by mugenggeng** [comment]
> Hi Thanks for your interest to our work. I think you test the GFLOPs for a batch of samples. The result we reported is done with batch size set to 1.
> 
> For your information, I include the code we used as below.
> 
> ```
> param_size = 0
> for param in model.parameters():
>     param_size += param.nelement() * param.element_size()
>     logger.info('Number of parameters: {}. Model Size: {:.2f} MB'.format(sum(p.numel() for p in model.parameters()), param_size/ 1024**2))
> 
> def prepare_input(resolution):
>         from rekognition_online_action_detection.models.feature_head import FEATURE_SIZES
>         fuse_dim = cfg.MODEL.LSTR.LONG_MEMORY_NUM_SAMPLES+cfg.MODEL.LSTR.WORK_MEMORY_NUM_SAMPLES
>         rgb = torch.FloatTensor(1, fuse_dim, FEATURE_SIZES[cfg.INPUT.VISUAL_FEATURE]).to(device)
>         flow = torch.FloatTensor(1, fuse_dim, FEATURE_SIZES[cfg.INPUT.MOTION_FEATURE]).to(device)
>         mask = torch.FloatTensor(1, cfg.MODEL.LSTR.LONG_MEMORY_NUM_SAMPLES).to(device)
>         return {'visual_inputs': rgb, 'motion_inputs': flow,  'memory_key_padding_mask': mask}
> 
> from ptflops import get_model_complexity_info
> macs, params = get_model_complexity_info(model, (1, 224, 224), as_strings=True, backend='pytorch',
>                                              input_constructor=prepare_input,
>                                              print_per_layer_stat = True, verbose = True)
> print('{:<30}  {:<8}'.format('Computational complexity: ', macs))
> print('{:<30}  {:<8}'.format('Number of parameters: ', params))
> ```

Thank you very much for your response. I have identified the problem. You calculated the GFLOPs when the batch size was 1. However, in fact, your accuracy was obtained through batch inference rather than streaming inference. Could I understand it this way? Furthermore, based on the code you provided, the GFLPOs I obtained are 2.95G, which is slightly different from that stated in the article. However, this is not significant. Finally, I would like to thank you for providing the GFLOPs calculation method. I have no other questions now.

**Comment by pangzhan27** [comment]
1. Taking batch or streaming inference does not affect the final performance as we make the prediction one frame per time during inference(only use the latest frame in the short-term memory), but streaming inference will be faster(in FPS) since it uses cache to avoid recomputing some intermediate results.
3. Yes, for evaluation, we use batch inference to speed up. To simplify and also meet the streaming setting, we report the efficiency with batchsize=1. We also test other methods under the same setting to make fair comparison. 
4. I cannot recall the details, the code i provide gives MACS, not GFLOPS, you should double check this (If i am not wrong, you should time 2 to get GFLOPS).

**Comment by mugenggeng** [comment]
> 1. Taking batch or streaming inference does not affect the final performance as we make the prediction one frame per time during inference(only use the latest frame in the short-term memory), but streaming inference will be faster(in FPS) since it uses cache to avoid recomputing some intermediate results.
> 2. Yes, for evaluation, we use batch inference to speed up. To simplify and also meet the streaming setting, we report the efficiency with batchsize=1. We also test other methods under the same setting to make fair comparison.
> 3. I cannot recall the details, the code i provide gives MACS, not GFLOPS, you should double check this (If i am not wrong, you should time 2 to get GFLOPS).

Thank you for your reply. FLOPs represents the total number of floating-point operations, and MAC stands for one addition and one multiplication. Since both addition and multiplication fall under floating-point operations, it should be that MAC = GFLOPs / 2, right?

**Comment by pangzhan27** [comment]
Based on my understanding, yes.

**Comment by mugenggeng** [comment]
Ok, thanks for your response. Undeniably, this is a good paper with a very novel thoughts.

**Comment by mugenggeng** [comment]
I have one more small question. I noticed that some of your hyperparameters are different from MAT, such as the number of heads in the Transformer, the learning rate, and so on. Could you please tell me whether you set it based on your experience, or used a tool to find it, or did it through some other method? This question has nothing to do with the paper. If you are willing to answer it, I would be extremely grateful.

**Comment by pangzhan27** [comment]
We manually tuned the hyper parameters. Some, such as the number of heads, are less critical to overall performance. Others, like the learning rate, were adjusted based on specific design considerations. Since MAT involves multiple rounds of refinement, certain parameters need to be adapted according to the different structure design.

**Comment by mugenggeng** [comment]
Thank you for your reply. I have no other questions.
